File: android/src/main/kotlin/com/atomaxinc/skalekit/flutter/SkaleKitPlugin.kt
package com.atomaxinc.skalekit.flutter

import android.app.Activity
import android.bluetooth.BluetoothDevice
import android.content.Context
import androidx.annotation.NonNull
import com.atomaxinc.skalekit.SkaleDevicePicker
import com.atomaxinc.skalekit.SkaleKit
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

class SkaleKitPlugin : FlutterPlugin, MethodCallHandler, ActivityAware, SkaleKit.SkaleListener {

    private lateinit var methodChannel: MethodChannel
    private lateinit var weightChannel: EventChannel
    private lateinit var connectionStateChannel: EventChannel
    private lateinit var buttonChannel: EventChannel
    private lateinit var deviceChannel: EventChannel

    private var applicationContext: Context? = null
    private var activity: Activity? = null
    private var skaleKit: SkaleKit? = null
    private var devicePicker: SkaleDevicePicker? = null

    private var weightEventSink: EventChannel.EventSink? = null
    private var connectionStateEventSink: EventChannel.EventSink? = null
    private var buttonEventSink: EventChannel.EventSink? = null
    private var deviceEventSink: EventChannel.EventSink? = null

    private var pendingDevicePickerResult: Result? = null
    private var pendingBatteryResult: Result? = null

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = flutterPluginBinding.applicationContext

        methodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "com.atomaxinc.skalekit/methods")
        methodChannel.setMethodCallHandler(this)

        weightChannel = EventChannel(flutterPluginBinding.binaryMessenger, "com.atomaxinc.skalekit/weight")
        weightChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                weightEventSink = events
            }
            override fun onCancel(arguments: Any?) {
                weightEventSink = null
            }
        })

        connectionStateChannel = EventChannel(flutterPluginBinding.binaryMessenger, "com.atomaxinc.skalekit/connectionState")
        connectionStateChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                connectionStateEventSink = events
            }
            override fun onCancel(arguments: Any?) {
                connectionStateEventSink = null
            }
        })

        buttonChannel = EventChannel(flutterPluginBinding.binaryMessenger, "com.atomaxinc.skalekit/button")
        buttonChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                buttonEventSink = events
            }
            override fun onCancel(arguments: Any?) {
                buttonEventSink = null
            }
        })

        deviceChannel = EventChannel(flutterPluginBinding.binaryMessenger, "com.atomaxinc.skalekit/devices")
        deviceChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                deviceEventSink = events
            }
            override fun onCancel(arguments: Any?) {
                deviceEventSink = null
            }
        })
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        weightChannel.setStreamHandler(null)
        connectionStateChannel.setStreamHandler(null)
        buttonChannel.setStreamHandler(null)
        deviceChannel.setStreamHandler(null)
        skaleKit?.disconnect()
        applicationContext = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        initializeSkaleKit()
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
        initializeSkaleKit()
    }

    override fun onDetachedFromActivity() {
        skaleKit?.disconnect()
        activity = null
    }

    private fun initializeSkaleKit() {
        activity?.let { ctx ->
            if (skaleKit == null) {
                skaleKit = SkaleKit(ctx)
                skaleKit?.setListener(this)
            }
        }
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        when (call.method) {
            "isConnected" -> {
                val state = skaleKit?.let {
                    // Check if currently connected
                    true // Placeholder - actual implementation depends on SkaleKit API
                } ?: false
                result.success(false) // Return false as we can't directly check connection state
            }

            "isBluetoothEnabled" -> {
                result.success(skaleKit?.isBluetoothEnabled() ?: false)
            }

            "hasPermissions" -> {
                applicationContext?.let { ctx ->
                    result.success(SkaleKit.hasPermissions(ctx))
                } ?: result.success(false)
            }

            "requestPermissions" -> {
                // Permission requests need to be handled by the Flutter app using permission_handler
                applicationContext?.let { ctx ->
                    result.success(SkaleKit.hasPermissions(ctx))
                } ?: result.success(false)
            }

            "startScan" -> {
                skaleKit?.let {
                    if (!it.isBluetoothEnabled()) {
                        result.error("BLUETOOTH_DISABLED", "Bluetooth is not enabled", null)
                        return
                    }
                    applicationContext?.let { ctx ->
                        if (!SkaleKit.hasPermissions(ctx)) {
                            result.error("PERMISSION_DENIED", "Required permissions not granted", null)
                            return
                        }
                    }
                    it.startScan()
                    connectionStateEventSink?.success("SCANNING")
                    result.success(null)
                } ?: result.error("NOT_INITIALIZED", "SkaleKit not initialized", null)
            }

            "stopScan" -> {
                // SkaleKit doesn't have a direct stopScan method
                result.success(null)
            }

            "showDevicePicker" -> {
                showDevicePicker(result)
            }

            "connect" -> {
                // Direct connection is handled through device picker
                result.error("NOT_SUPPORTED", "Direct connection by ID is not supported. Use showDevicePicker instead.", null)
            }

            "disconnect" -> {
                skaleKit?.disconnect()
                result.success(null)
            }

            "tare" -> {
                skaleKit?.tare()
                result.success(null)
            }

            "getBatteryLevel" -> {
                skaleKit?.let {
                    pendingBatteryResult = result
                    it.requestBatteryLevel()
                } ?: result.error("NOT_INITIALIZED", "SkaleKit not initialized", null)
            }

            "setLEDDisplay" -> {
                // LED display control may not be available in Android SDK
                result.success(null)
            }

            "setAutoConnect" -> {
                // Auto-connect feature implementation depends on SkaleKit API
                result.success(null)
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    private fun showDevicePicker(result: Result) {
        val currentActivity = activity
        val kit = skaleKit

        if (currentActivity == null || kit == null) {
            result.error("NOT_INITIALIZED", "Activity or SkaleKit not initialized", null)
            return
        }

        if (!kit.isBluetoothEnabled()) {
            result.error("BLUETOOTH_DISABLED", "Bluetooth is not enabled", null)
            return
        }

        applicationContext?.let { ctx ->
            if (!SkaleKit.hasPermissions(ctx)) {
                result.error("PERMISSION_DENIED", "Required permissions not granted", null)
                return
            }
        }

        pendingDevicePickerResult = result
        connectionStateEventSink?.success("SCANNING")

        devicePicker = SkaleDevicePicker(currentActivity, kit)
        devicePicker?.show(object : SkaleDevicePicker.OnDeviceSelectedListener {
            override fun onDeviceSelected(device: BluetoothDevice) {
                connectionStateEventSink?.success("CONNECTING")
                kit.connect(device)
                // The result will be sent when onConnectionStateChanged is called
            }

            override fun onCancelled() {
                connectionStateEventSink?.success("DISCONNECTED")
                pendingDevicePickerResult?.error("CANCELLED", "User cancelled device selection", null)
                pendingDevicePickerResult = null
            }
        })
    }

    // SkaleKit.SkaleListener implementation

    override fun onConnectionStateChanged(state: SkaleKit.ConnectionState) {
        activity?.runOnUiThread {
            val stateString = when (state) {
                SkaleKit.ConnectionState.DISCONNECTED -> "DISCONNECTED"
                SkaleKit.ConnectionState.SCANNING -> "SCANNING"
                SkaleKit.ConnectionState.CONNECTING -> "CONNECTING"
                SkaleKit.ConnectionState.CONNECTED -> "CONNECTED"
            }
            connectionStateEventSink?.success(stateString)

            if (state == SkaleKit.ConnectionState.CONNECTED) {
                pendingDevicePickerResult?.success(null)
                pendingDevicePickerResult = null
            }
        }
    }

    override fun onWeightUpdate(weight: Float) {
        activity?.runOnUiThread {
            weightEventSink?.success(weight.toDouble())
        }
    }

    override fun onButtonClicked(buttonId: Int) {
        activity?.runOnUiThread {
            buttonEventSink?.success(buttonId)
        }
    }

    override fun onBatteryLevelUpdate(level: Int) {
        activity?.runOnUiThread {
            pendingBatteryResult?.success(level)
            pendingBatteryResult = null
        }
    }

    override fun onError(error: SkaleKit.SkaleError) {
        activity?.runOnUiThread {
            val errorCode = when (error) {
                is SkaleKit.SkaleError.BluetoothDisabled -> "BLUETOOTH_DISABLED"
                is SkaleKit.SkaleError.PermissionDenied -> "PERMISSION_DENIED"
                is SkaleKit.SkaleError.DeviceNotFound -> "DEVICE_NOT_FOUND"
                is SkaleKit.SkaleError.ConnectionFailed -> "CONNECTION_FAILED"
                is SkaleKit.SkaleError.ConnectionLost -> "CONNECTION_LOST"
                is SkaleKit.SkaleError.Unknown -> "UNKNOWN"
            }
            val errorMessage = when (error) {
                is SkaleKit.SkaleError.Unknown -> error.message ?: "Unknown error"
                else -> errorCode
            }

            connectionStateEventSink?.error(errorCode, errorMessage, null)

            pendingDevicePickerResult?.error(errorCode, errorMessage, null)
            pendingDevicePickerResult = null
        }
    }
}

File: ios/Classes/SkaleKitPlugin.swift
import Flutter
import UIKit
import SkaleKit
import CoreBluetooth

public class SkaleKitPlugin: NSObject, FlutterPlugin, SKSkaleDelegate {

    private let skale = SKSkale()
    private var methodChannel: FlutterMethodChannel?
    private var weightEventSink: FlutterEventSink?
    private var connectionStateEventSink: FlutterEventSink?
    private var buttonEventSink: FlutterEventSink?
    private var deviceEventSink: FlutterEventSink?

    private var centralManager: CBCentralManager?
    private var discoveredDevices: [String: [String: Any]] = [:]
    private var isScanning = false

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = SkaleKitPlugin()

        // Method channel
        let methodChannel = FlutterMethodChannel(
            name: "com.atomaxinc.skalekit/methods",
            binaryMessenger: registrar.messenger()
        )
        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        instance.methodChannel = methodChannel

        // Weight event channel
        let weightChannel = FlutterEventChannel(
            name: "com.atomaxinc.skalekit/weight",
            binaryMessenger: registrar.messenger()
        )
        weightChannel.setStreamHandler(WeightStreamHandler(plugin: instance))

        // Connection state event channel
        let connectionStateChannel = FlutterEventChannel(
            name: "com.atomaxinc.skalekit/connectionState",
            binaryMessenger: registrar.messenger()
        )
        connectionStateChannel.setStreamHandler(ConnectionStateStreamHandler(plugin: instance))

        // Button event channel
        let buttonChannel = FlutterEventChannel(
            name: "com.atomaxinc.skalekit/button",
            binaryMessenger: registrar.messenger()
        )
        buttonChannel.setStreamHandler(ButtonStreamHandler(plugin: instance))

        // Device discovery event channel
        let deviceChannel = FlutterEventChannel(
            name: "com.atomaxinc.skalekit/devices",
            binaryMessenger: registrar.messenger()
        )
        deviceChannel.setStreamHandler(DeviceStreamHandler(plugin: instance))

        instance.skale.delegate = instance
        instance.skale.isAutoConnectEnabled = false
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "isConnected":
            result(skale.isConnected)

        case "isBluetoothEnabled":
            if centralManager == nil {
                centralManager = CBCentralManager(delegate: nil, queue: nil, options: [CBCentralManagerOptionShowPowerAlertKey: false])
            }
            result(centralManager?.state == .poweredOn)

        case "hasPermissions":
            if #available(iOS 13.1, *) {
                let authorization = CBCentralManager.authorization
                result(authorization == .allowedAlways || authorization == .restricted)
            } else {
                result(true)
            }

        case "requestPermissions":
            // On iOS, permissions are requested automatically when CBCentralManager is created
            if centralManager == nil {
                centralManager = CBCentralManager(delegate: nil, queue: nil)
            }
            if #available(iOS 13.1, *) {
                let authorization = CBCentralManager.authorization
                result(authorization == .allowedAlways || authorization == .restricted)
            } else {
                result(true)
            }

        case "startScan":
            startScanning(result: result)

        case "stopScan":
            stopScanning()
            result(nil)

        case "showDevicePicker":
            showDevicePicker(result: result)

        case "connect":
            if let args = call.arguments as? [String: Any],
               let deviceId = args["id"] as? String {
                connect(deviceId: deviceId, result: result)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Device ID is required", details: nil))
            }

        case "disconnect":
            skale.disconnect()
            result(nil)

        case "tare":
            skale.tare()
            result(nil)

        case "getBatteryLevel":
            skale.readBatteryLife { level in
                result(Int(level))
            }

        case "setLEDDisplay":
            if let args = call.arguments as? [String: Any],
               let isOn = args["isOn"] as? Bool {
                skale.setLEDDisplayOn(isOn)
                result(nil)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "isOn parameter is required", details: nil))
            }

        case "setAutoConnect":
            if let args = call.arguments as? [String: Any],
               let enabled = args["enabled"] as? Bool {
                skale.isAutoConnectEnabled = enabled
                result(nil)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "enabled parameter is required", details: nil))
            }

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Scanning

    private func startScanning(result: @escaping FlutterResult) {
        guard let manager = centralManager, manager.state == .poweredOn else {
            if centralManager == nil {
                centralManager = CBCentralManager(delegate: nil, queue: nil)
            }
            result(FlutterError(code: "BLUETOOTH_DISABLED", message: "Bluetooth is not enabled", details: nil))
            return
        }

        discoveredDevices.removeAll()
        isScanning = true
        connectionStateEventSink?("SCANNING")

        // Note: The actual scanning is handled by the native SDK
        // This is a placeholder for custom scanning if needed
        result(nil)
    }

    private func stopScanning() {
        isScanning = false
    }

    // MARK: - Device Picker

    private func showDevicePicker(result: @escaping FlutterResult) {
        // Must run on main thread for UI
        DispatchQueue.main.async { [weak self] in
            guard let self = self else {
                result(FlutterError(code: "PLUGIN_ERROR", message: "Plugin instance deallocated", details: nil))
                return
            }

            guard let topController = self.getTopViewController() else {
                result(FlutterError(code: "NO_VIEW_CONTROLLER", message: "Could not find root view controller", details: nil))
                return
            }

            self.connectionStateEventSink?("SCANNING")

            self.skale.showDevicePicker(onViewContoller: topController) { [weak self] error in
                if let error = error {
                    let nsError = error as NSError
                    let flutterError = self?.mapNSErrorToFlutterError(nsError)
                    result(flutterError)
                } else {
                    // Connection will be handled by delegate methods
                    // Return nil to indicate picker was shown successfully
                    result(nil)
                }
            }
        }
    }

    private func getTopViewController() -> UIViewController? {
        var rootViewController: UIViewController?

        if #available(iOS 13.0, *) {
            // iOS 13+ use scene-based approach
            let scenes = UIApplication.shared.connectedScenes
            let windowScene = scenes.first as? UIWindowScene
            rootViewController = windowScene?.windows.first(where: { $0.isKeyWindow })?.rootViewController
        } else {
            rootViewController = UIApplication.shared.keyWindow?.rootViewController
        }

        guard var topController = rootViewController else {
            return nil
        }

        // Find the topmost presented view controller
        while let presented = topController.presentedViewController {
            topController = presented
        }

        return topController
    }

    private func connect(deviceId: String, result: @escaping FlutterResult) {
        // The native SDK uses the device picker for connection
        // Direct connection by ID is not supported in the current API
        // We'll use the device picker flow instead
        result(FlutterError(code: "NOT_SUPPORTED", message: "Direct connection by ID is not supported. Use showDevicePicker instead.", details: nil))
    }

    // MARK: - SKSkaleDelegate

    public func skaleDidConnected(_ skale: SKSkale) {
        connectionStateEventSink?("CONNECTED")
    }

    public func skaleDidDisconnected(_ skale: SKSkale) {
        connectionStateEventSink?("DISCONNECTED")
    }

    public func skale(_ skale: SKSkale, didErrorOccur error: Error) {
        let nsError = error as NSError
        connectionStateEventSink?(FlutterError(
            code: mapErrorCode(nsError.code),
            message: nsError.localizedDescription,
            details: nil
        ))
    }

    public func skaleWeightDidUpdate(_ weight: Float) {
        weightEventSink?(Double(weight))
    }

    public func skaleDeviceDidClickButton(_ buttonIndex: UInt) {
        buttonEventSink?(Int(buttonIndex))
    }

    // MARK: - Error Mapping

    private func mapNSErrorToFlutterError(_ error: NSError) -> FlutterError {
        let code = mapErrorCode(error.code)
        return FlutterError(code: code, message: error.localizedDescription, details: nil)
    }

    private func mapErrorCode(_ code: Int) -> String {
        switch code {
        case 100: // SKErrorLECancel
            return "CANCELLED"
        case 101: // SKErrorLENotAvailable
            return "BLUETOOTH_DISABLED"
        case 102: // SKErrorLEDeviceAlreadyConnected
            return "ALREADY_CONNECTED"
        case 103: // SKErrorLEConnectionUnknown
            return "CONNECTION_FAILED"
        case 104: // SKErrorLEConnectionInvalidParameters
            return "CONNECTION_FAILED"
        case 105: // SKErrorLEConnectionInvalidHandle
            return "CONNECTION_FAILED"
        case 106: // SKErrorLEConnectionTimeout
            return "CONNECTION_TIMEOUT"
        case 107: // SKErrorLEValidation
            return "CONNECTION_FAILED"
        case 108: // SKErrorLEInvalidDevice
            return "DEVICE_NOT_FOUND"
        case 109: // SKErrorLEAccess
            return "PERMISSION_DENIED"
        default:
            return "UNKNOWN"
        }
    }

    // MARK: - Event Sink Setters

    func setWeightEventSink(_ sink: FlutterEventSink?) {
        weightEventSink = sink
    }

    func setConnectionStateEventSink(_ sink: FlutterEventSink?) {
        connectionStateEventSink = sink
    }

    func setButtonEventSink(_ sink: FlutterEventSink?) {
        buttonEventSink = sink
    }

    func setDeviceEventSink(_ sink: FlutterEventSink?) {
        deviceEventSink = sink
    }
}

// MARK: - Stream Handlers

class WeightStreamHandler: NSObject, FlutterStreamHandler {
    weak var plugin: SkaleKitPlugin?

    init(plugin: SkaleKitPlugin) {
        self.plugin = plugin
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        plugin?.setWeightEventSink(events)
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        plugin?.setWeightEventSink(nil)
        return nil
    }
}

class ConnectionStateStreamHandler: NSObject, FlutterStreamHandler {
    weak var plugin: SkaleKitPlugin?

    init(plugin: SkaleKitPlugin) {
        self.plugin = plugin
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        plugin?.setConnectionStateEventSink(events)
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        plugin?.setConnectionStateEventSink(nil)
        return nil
    }
}

class ButtonStreamHandler: NSObject, FlutterStreamHandler {
    weak var plugin: SkaleKitPlugin?

    init(plugin: SkaleKitPlugin) {
        self.plugin = plugin
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        plugin?.setButtonEventSink(events)
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        plugin?.setButtonEventSink(nil)
        return nil
    }
}

class DeviceStreamHandler: NSObject, FlutterStreamHandler {
    weak var plugin: SkaleKitPlugin?

    init(plugin: SkaleKitPlugin) {
        self.plugin = plugin
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        plugin?.setDeviceEventSink(events)
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        plugin?.setDeviceEventSink(nil)
        return nil
    }
}
